File: src/main/cspom/variable/ExtensiveDomain.java
/**
 * CSPFJ Competitor - CSP solver using the CSPFJ API for Java
 * Copyright (C) 2006 Julien VION
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

package cspom.variable;

import java.util.ArrayList;
import java.util.List;

public final class ExtensiveDomain implements Domain {

	private final List<Number> values;

	private final String name;
	
	private final int nbValues;

	public ExtensiveDomain(final String name, final List<Number> values) {
		super();
		this.values = new ArrayList<Number>(values);
		this.name = name;
		this.nbValues = values.size();
	}

	public List<Number> getValues() {
		return values;
	}

	public String getName() {
		return name;
	}

	public int getNbValues() {
		return nbValues;
	}
	
	@Override
	public String toString() {
		final StringBuilder stb = new StringBuilder(name);
		stb.append(": [").append(values.size()).append(']');
		// for (Number value : values) {
		// stb.append(value).append(", ");
		// }
		// stb.delete(stb.length() - 2, stb.length()).append(']');
		return stb.toString();
	}

}

File: src/main/scala/cspom/variable/ExtensiveDomain.scala
package cspom.variable

class ExtensiveDomain[T](val values: Seq[T]) extends CSPOMDomain[T] {
  override val hashCode = values.hashCode

  override def equals(obj: Any) = obj match {
    case ed: ExtensiveDomain[T] => ed.values == values
    case _ => false
  }

  override def toString =
    if (size > 5)
      values.take(5).mkString("{", ", ", "...}")
    else
      values.mkString("{", ", ", "}")

  def toXCSP = values.mkString(", ")

  def intersect[S >: T](domain: CSPOMDomain[S]) =
    new ExtensiveDomain[S](this.values.intersect(domain.values))
}

object ExtensiveDomain {
  def of[T](values: T*) = new ExtensiveDomain(values)
}